proposals/0031-hlsl-vector-matrix-operations.md

```diff
@@ -160,7 +160,58 @@ TBD
 
 ### Builtins
 
-TBD
+Although these "builtins" are not intended to be part of the HLSL language, and
+no promises are made that these will continue to be available over time, this
+proposal describes an implementation in terms of builtins such as this. For
+this reason it is useful to have them described here as a reference point.
+
+Each builtin corresponds to one of the operations described in [0029].
+
+```c++
+namespace dx {
+namespace linalg {
+namespace details {
+
+// dx.op.matvecmul
+template <typename TYo, int NUMo, typename TYi, int NUMi, typename RES>
+void __builtin_MatVecMul(vector<TYi, NUMi> InputVector,
+                         uint InputVectorInterpretation, RES MatrixResource,
+                         uint MatrixStartOffset, uint MatrixInterpretation,
+                         uint M, uint K, uint Layout, bool MatrixTranspose,
+                         uint MatrixStride, out vector<TYo, NUMo> OutputVector);
+
+// dx.op.matvecmuladd
+template <typename TYo, int NUMo, typename TYi, int NUMi, typename RESm,
+          typename RESv>
+void __builtin_MatVecMulAdd(vector<TYi, NUMi> InputVector,
+                            uint InputVectorInterpretation, RESm MatrixResource,
+                            uint MatrixStartOffset, uint MatrixInterpretation,
+                            uint M, uint K, uint Layout, bool MatrixTranspose,
+                            uint MatrixStride, RESv BiasVectorResource,
+                            uint BiasVectorOffset,
+                            uint BiasVectorInterpretation,
+                            out vector<TYo, NUMo> OutputVector);
+
+// dx.op.outerproductaccumulate
+template <typename TY, int M, int N, typename RES>
+void __builtin_OuterProductAccumulate(vector<TY, M> InputVector1,
+                                      vector<TY, N> InputVector2,
+                                      RES MatrixResource,
+                                      uint MatrixStartOffset,
+                                      DataType MatrixInterpretation,
```

Comment:
should MatrixInterpretation and Layout be uints here or do we want enums?

Reply:
I think they need to be uints - I'm trying to define these builtins with the minimum number of dependencies, so I don't want to also have to define the enums for them.